clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslServerAuthenticator.java

```diff
@@ -215,6 +217,9 @@ public void authenticate() throws IOException {
             try {
                 switch (saslState) {
                     case HANDSHAKE_REQUEST:
+                        handleKafkaRequest(clientToken);
+                        break;
+                    case GSSAPI_OR_HANDSHAKE_REQUEST:
```

Comment:
@junrao We are relying on fall-through for the second case, so it's not straightforward to make that change.